fdb-record-layer-spatial/src/main/java/com/apple/foundationdb/record/spatial/geophile/GeophileSpatialFunctionKeyExpression.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * SpatialFunctionKeyExpression.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.spatial.geophile;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.metadata.Key;
+import com.apple.foundationdb.record.metadata.expressions.FunctionKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecord;
+import com.geophile.z.Space;
+import com.geophile.z.SpatialObject;
+import com.geophile.z.spatialobject.d2.Point;
+import com.google.protobuf.Message;
+import org.locationtech.jts.io.ParseException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Spatial function key expressions.
+ */
+@API(API.Status.EXPERIMENTAL)
+public abstract class GeophileSpatialFunctionKeyExpression extends FunctionKeyExpression {
+    public static final Space SPACE_LAT_LON = GeophileSpatial.createLatLonSpace();
+
+    private final Space space;
+
+    protected GeophileSpatialFunctionKeyExpression(@Nonnull String name, @Nonnull KeyExpression arguments) {
+        super(name, arguments);
+        // TODO: How do we make this a part of the key expression? A naming convention?
+        this.space = SPACE_LAT_LON;
```

Comment:
Having "geo" in the name imply `SPACE_LAT_LON` seems right. This coordinate system, including all its quirks with wraparound, is pretty much fixed.

But for other coordinate systems, there is a lot more flexibility. Since longs are used for Z values, there is a limit of fifty-odd bits, but they can be pretty much distributed how you want. So I think you'd need a name registry where you define them. I don't think it'd be practical to literally parse name to get numbers for the ranges, number of bits and so on.

Reply:
Definitely agree that we should not be parsing the name to determine parameters. I think a registry system is probably good enough for the more complicated spaces. Perhaps @nschiefer could weigh in as well, but it might also be something that we should think about more when we add the other spaces.